### PR TITLE
[SYCL] Implementation of preprocessing output when option '-include-footer' is used.

### PR DESCRIPTION
**Pull request by zahiraam**
Current implementation to support our integration header and footer files during the host specific compilation, is to create a temporary file based on the source file with the integration footer appended to it. This temporary file is then compiled as the main source file.
Use of `ccache` doesn't work well with this implementation, as the actual compiled source file does not match the source file given on the command line.
To solve this, a new option named `-include-footer` is introduced. With it, the integration footer is appended to the source file during preprocessing. 

Syntax : `-include-footer filename`
`filename`: represents the integration footer file name generated during the device compilation.

Use: `clang -cc1 -include int-header.h -include-footer int-footer.h main.cpp`

This patch implements the changes produced in the output of the preprocessing phase.  It only uses this option. The implementation of the option will be done in a different patch. 
The preprocessed output should include the name of the footer and its content at the end of the source code.

### CODE REVIEW

**Review by mdtoguchi — COMMENTED** [review]
any changes in Driver.cpp or Clang.cpp shouldn't be required for your functionality.  I can integrate those changes along with the proper test updates after your PR is merged.

File: clang/include/clang/Driver/Options.td [hunk 1]
```diff
@@ -4706,6 +4706,10 @@ def image__base : Separate<["-"], "image_base">;
 def include_ : JoinedOrSeparate<["-", "--"], "include">, Group<clang_i_Group>, EnumName<"include">,
     MetaVarName<"<file>">, HelpText<"Include file before parsing">,
     Visibility<[ClangOption, CC1Option]>;
+def include_footer : Separate<["-"], "include-footer">, Group<clang_i_Group>,
+    Visibility<[ClangOption, CC1Option]>,
```
Comment (mdtoguchi) [inline clang/include/clang/Driver/Options.td:4710]: There is no intention of having this option be visible to the end user due to the limited usage model, I think `CC1Option` only should be OK.

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 2]
```diff
@@ -5584,6 +5584,19 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
...
+      if (types::getPreprocessedType(Input.getType()) != types::TY_INVALID &&
+          !Args.hasArg(options::OPT_fno_sycl_use_footer) && !Footer.empty()) {
+        CmdArgs.push_back("-include-footer");
+        CmdArgs.push_back(Args.MakeArgString(Footer));
+        // When creating dependency information, filter out the generated
+        // integration footer file.
+        CmdArgs.push_back("-dependency-filter");
+        CmdArgs.push_back(Args.MakeArgString(Footer));
+      }
+
```
Comment (mdtoguchi) [inline clang/lib/Driver/ToolChains/Clang.cpp:5599]: This and the changes to `test/Driver/sycl-int-footer.cpp` should be taken care of in a separate PR that will cover the entirety of the switch to using `-include-footer` instead of `append-file`.  Performing this now will cause the footer to be added twice.

Comment (zahiraam) [reply clang/lib/Driver/ToolChains/Clang.cpp:5599]: OK. I will remove this code and the test from the PR.

**Review by mdtoguchi — APPROVED** [review]
OK for driver

File: clang/lib/Frontend/PrintPreprocessedOutput.cpp [hunk 3]
```diff
@@ -806,6 +837,40 @@ struct UnknownPragmaHandler : public PragmaHandler {
...
+  SourceManager &SourceMgr = PP.getSourceManager();
+  PresumedLoc UserLoc = SourceMgr.getPresumedLoc(Loc);
+  if (UserLoc.isInvalid())
+    return;
+  FileID FooterFileID = ComputeValidFooterFileID(SourceMgr, Footer);
+  StringRef FooterContentBuffer = SourceMgr.getBufferData(FooterFileID);
+  unsigned NewLine = UserLoc.getLine();
+  // print out the name of the integration footer.
+  Callbacks->WriteFooterInfo(NewLine, Footer);
+  std::string StrippedFooterContentBuffer;
```
Comment (premanandrao) [inline clang/lib/Frontend/PrintPreprocessedOutput.cpp:864]: This is unused.

File: clang/lib/Frontend/PrintPreprocessedOutput.cpp [hunk 4]
```diff
@@ -806,6 +837,40 @@ struct UnknownPragmaHandler : public PragmaHandler {
...
+  SmallVector<StringRef, 8> FooterContentArr;
+  FooterContentBuffer.split(FooterContentArr, '\r');
```
Comment (premanandrao) [inline clang/lib/Frontend/PrintPreprocessedOutput.cpp:866]: Can you explain what this split with carriage-return does?   For my curiosity.  Is this done elsewhere that I could see?

File: clang/lib/Frontend/PrintPreprocessedOutput.cpp [hunk 5]
```diff
@@ -324,6 +343,18 @@ void PrintPPOutputPPCallbacks::startNewLineIfNeeded() {
     EmittedDirectiveOnThisLine = false;
   }
 }
+void PrintPPOutputPPCallbacks::AddFooter(SourceLocation Loc,
+                                         std::string Footer) {
+  SourceManager &SourceMgr = SM;
```
Comment (premanandrao) [inline clang/lib/Frontend/PrintPreprocessedOutput.cpp:348]: Why is this needed?   Can't you use SM directly below?

File: clang/test/Preprocessor/Inputs/file2.h [hunk 6]
```diff
@@ -0,0 +1,5 @@
+int file2() {
+  return 2;
+}
+
```
Comment (premanandrao) [inline clang/test/Preprocessor/Inputs/file2.h:4]:
```suggestion
```

File: clang/test/Preprocessor/sycl-int-footer.cpp [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+// CHECK: # 2 "<built-in>" 2
+// CHECK: # 1 "[[INPUTFILE:.+\.cpp]]" 2
+// CHECK: int main() {
+// CHECK:   int i = 0;
+// CHECK:   return i++;
+// CHECK: }
+// CHECK: # 1 "[[INTFOOTER:.+\.h]]" 1
+// CHECK: int file2() {
+// CHECK:   return 2;
+// CHECK: }
```
Comment (premanandrao) [inline clang/test/Preprocessor/sycl-int-footer.cpp:24]: Is the header file and footer file included in device compilations too now?

File: clang/lib/Frontend/PrintPreprocessedOutput.cpp (hunk 4 above)
Comment (zahiraam) [reply clang/lib/Frontend/PrintPreprocessedOutput.cpp:866]: split takes a string a returns an array of strings where each element of an array is the content between two separators, in this case \r is the separator.

File: clang/test/Preprocessor/sycl-int-footer.cpp (hunk 7 above)
Comment (mdtoguchi) [reply clang/test/Preprocessor/sycl-int-footer.cpp:24]: The integration header and integration footer are only used in the host compilation.  Device compilations are when the header/footer are created.

Comment (mdtoguchi) [reply clang/test/Preprocessor/sycl-int-footer.cpp:24]: I can see where the confusion is - adding `-fsycl` here implies we are adding the header and footer to both the host and device compilations.  I think just a regular host compilation without `-fsycl` to test should be sufficient.

Comment (zahiraam) [reply clang/test/Preprocessor/sycl-int-footer.cpp:22]: Since the `-include-footer` will be used only in host compilation, I have updated the LIT test to reflect that.

File: clang/lib/Frontend/PrintPreprocessedOutput.cpp (hunk 4 above)
Comment (premanandrao) [reply clang/lib/Frontend/PrintPreprocessedOutput.cpp:859]: I should have been clearer.   I understand how the function works generally, but was wondering more about where the '\r' gets introduced from.   Do we add it explicitly somewhere else?

**Review by mdtoguchi — APPROVED** [review]

File: clang/lib/Frontend/PrintPreprocessedOutput.cpp (hunk 4 above)
Comment (zahiraam) [reply clang/lib/Frontend/PrintPreprocessedOutput.cpp:859]: Sorry I misunderstood your question.
I am not quite sure who adds those. I am assuming that the content of the footer would be some regular C++ code. When I enter artificially in the footer code such that: 
`int foo() {
  int i
    return i++;
}
`
the `FooterContentBuffer` is a `StringRef` where end of lines are represented by `\r\n`.
Do you have by any chance a test case where the footer is actual code, to see concretely in what form the code footer would be?

Comment (premanandrao) [reply clang/lib/Frontend/PrintPreprocessedOutput.cpp:859]: Thanks.  `\r\n` was what I was expecting.   If we split with `\r` won't the part after the split have a `\n`?

For test cases, you could look at: clang/test/CodeGenSYCL/integration-footer.cpp or for a full test, look at: sycl/test-e2e/SpecConstants/2020/non_native/SpecConstBuffer.cpp

Comment (zahiraam) [reply clang/lib/Frontend/PrintPreprocessedOutput.cpp:859]: Yes, it will have.

Comment (zahiraam) [reply clang/lib/Frontend/PrintPreprocessedOutput.cpp:859]: Splitting at \r\n didn't work. So I will leave it as is if you agree.

Comment (premanandrao) [reply clang/lib/Frontend/PrintPreprocessedOutput.cpp:850]: Thanks for checking.   I am fine with leaving it as is.   I was just curious initially.

File: clang/lib/Frontend/PrintPreprocessedOutput.cpp [hunk 8]
```diff
@@ -806,6 +822,39 @@ struct UnknownPragmaHandler : public PragmaHandler {
...
+                               PrintPPOutputPPCallbacks *Callbacks) {
+  SourceManager &SourceMgr = PP.getSourceManager();
+  PresumedLoc UserLoc = SourceMgr.getPresumedLoc(Loc);
+  if (UserLoc.isInvalid())
+    return;
+  FileID FooterFileID = ComputeValidFooterFileID(SourceMgr, Footer);
+  StringRef FooterContentBuffer = SourceMgr.getBufferData(FooterFileID);
+  // print out the name of the integration footer.
+  Callbacks->WriteFooterInfo(Footer);
+  std::string StrippedFooterContentBuffer;
```
Comment (premanandrao) [inline clang/lib/Frontend/PrintPreprocessedOutput.cpp:848]:
```suggestion
```

**Review by premanandrao — COMMENTED** [review]
Just one unused variable still.

**Review by premanandrao — APPROVED** [review]

**Comment by zahiraam** [comment]
Thanks @mdtoguchi and @premanandrao.

**Comment by zahiraam** [comment]
@intel/llvm-gatekeepers can this PR be merged in please? Thanks.
